### PR TITLE
Fset 656 - Adjustment Management

### PR DESCRIPTION
**Pull request by kjharkin**

### CODE REVIEW

File: app/model/Commands.scala [hunk 1]
```diff
@@ -229,7 +238,8 @@ object Commands {
     implicit val candidateFormat = Json.format[Candidate]
     implicit val reportFormat = Json.format[Report]
     implicit val preferencesWithContactDetailsFormat = Json.format[PreferencesWithContactDetails]
-    implicit val adjustmentManagementFormat = Json.format[AdjustmentManagement]
+    implicit val adjustmentDetailFormat = Json.format[AdjustmentDetail]
+    implicit val adjustmentManagementFormatNew = Json.format[AdjustmentManagement]
```
Comment (henricook) [inline app/model/Commands.scala]: I guess we don't need the 'new' suffix any more

File: app/repositories/onlinetesting/Phase2TestRepository.scala [hunk 2]
```diff
@@ -83,11 +83,36 @@ class Phase2TestMongoRepository(dateTime: DateTimeFactory)(implicit mongo: () =>
...
+        BSONDocument("applicationStatus" -> ApplicationStatus.PHASE1_TESTS_PASSED),
+        BSONDocument(s"progress-status.${PHASE1_TESTS_PASSED}" -> true),
+        BSONDocument("$or" -> BSONArray(
+          BSONDocument("$and" -> BSONArray(
+            BSONDocument("assistance-details.needsSupportForOnlineAssessment" -> false),
+            BSONDocument("assistance-details.needsSupportAtVenue" -> false),
+            BSONDocument("assistance-details.guaranteedInterview" -> BSONDocument("$ne" -> true)))),
+          BSONDocument("$and" -> BSONArray(
+            BSONDocument("$or" -> BSONArray(
+              BSONDocument("assistance-details.guaranteedInterview" -> true),
```
Comment (ollyw) [inline app/repositories/onlinetesting/Phase2TestRepository.scala]: Is this needed `assistance-details.guaranteedInterview`? Doesn't the above logic invite candidates who don't need support without confirmation, so therefore redundant? I am trying to get my head around it

File: app/repositories/application/GeneralApplicationRepository.scala [hunk 3]
```diff
@@ -890,7 +889,7 @@ class GeneralApplicationMongoRepository(timeZoneService: TimeZoneService,
   def extract(key: String)(root: Option[BSONDocument]) = root.flatMap(_.getAs[String](key))
 
   /*private def getAdjustmentsConfirmed(assistance: Option[BSONDocument]): Option[String] = {
```
Comment (henricook) [inline app/repositories/application/GeneralApplicationRepository.scala]: Is this commented fasttrack legacy? Do you think we can ditch the whole method now?

File: app/repositories/onlinetesting/Phase2TestRepository.scala [hunk 4]
```diff
@@ -83,11 +83,36 @@ class Phase2TestMongoRepository(dateTime: DateTimeFactory)(implicit mongo: () =>
   }
 
   override def nextApplicationsReadyForOnlineTesting: Future[List[OnlineTestApplication]] = {
-    val query = BSONDocument(
-      "applicationStatus" -> ApplicationStatus.PHASE1_TESTS_PASSED,
-      "assistance-details.needsSupportForOnlineAssessment" -> false,
-      s"progress-status.${PHASE1_TESTS_PASSED}" -> true
-    )
+    // new
```
Comment (henricook) [inline app/repositories/onlinetesting/Phase2TestRepository.scala]: Probably don't need this comment any more?

File: app/repositories/onlinetesting/Phase2TestRepository.scala [hunk 5]
```diff
@@ -83,11 +83,36 @@ class Phase2TestMongoRepository(dateTime: DateTimeFactory)(implicit mongo: () =>
...
+    val query =
+      BSONDocument("$and" -> BSONArray(
+        BSONDocument("applicationStatus" -> ApplicationStatus.PHASE1_TESTS_PASSED),
+        BSONDocument(s"progress-status.${PHASE1_TESTS_PASSED}" -> true),
```
Comment (henricook) [inline app/repositories/onlinetesting/Phase2TestRepository.scala]: They were here before, but i don't think the curly braces are necessary here

File: app/services/application/ApplicationService.scala [hunk 6]
```diff
@@ -71,4 +72,47 @@ trait ApplicationService extends EventSink {
...
+
+    val standardEventList = DataStoreEvents.ManageAdjustmentsUpdated(applicationId) ::
+      AuditEvents.AdjustmentsConfirmed(Map("applicationId" -> applicationId, "adjustments" -> adjustmentInformation.toString)) ::
+      Nil
+
+    def toEmailString(header: String, adjustmentDetail: Option[AdjustmentDetail]): String ={
+      def mkString(ad: Option[AdjustmentDetail]): Option[String] = {
+        ad.map { e =>
+          val et = e.timeNeeded.map( tn => s"$tn% extra time")
+          val iv_info = e.invigilatedInfo
```
Comment (ollyw) [inline app/services/application/ApplicationService.scala]: what is going on with the naming conventions? Somebody been programming Ruby recently?

File: app/services/application/ApplicationService.scala [hunk 7]
```diff
@@ -71,4 +72,47 @@ trait ApplicationService extends EventSink {
...
+          val other_info = e.otherInfo
+          List(et,iv_info,other_info).flatten.mkString(", ")
```
Comment (ollyw) [inline app/services/application/ApplicationService.scala]: Spacing?

Comment (ollyw) [inline app/services/application/ApplicationService.scala]: Are these extra vals even making this more readable?

File: app/services/onlinetesting/Phase2TestService.scala [hunk 8]
```diff
@@ -169,20 +172,27 @@ trait Phase2TestService extends OnlineTestService with Phase2TestConcern with Sc
...
+          }).flatten
+        }
       }
     }
+  }
 
   private def registerAndInviteForTestGroup(applications: List[OnlineTestApplication], schedule: Phase2Schedule)
-    (implicit hc: HeaderCarrier, rh: RequestHeader): Future[List[OnlineTestApplication]] = applications match {
+                                           (implicit hc: HeaderCarrier, rh: RequestHeader): Future[List[OnlineTestApplication]]
+  = applications match {
```
Comment (henricook) [inline app/services/onlinetesting/Phase2TestService.scala]: I feel like the equals is a bit weird on this line, obviously it's valid - best argument i have is that it's inconsistent with the rest of the codebase

File: app/services/onlinetesting/Phase2TestService.scala [hunk 9]
```diff
@@ -296,18 +307,19 @@ trait Phase2TestService extends OnlineTestService with Phase2TestConcern with Sc
...
   }
 
-  //TODO Once the time adjustments ticket has been done then this should be updated to apply the etray adjustment settings.
-  def buildTimeAdjustments(needsAdjustment: Boolean, assessmentId: Int) = if (needsAdjustment) {
-    List(TimeAdjustments(assessmentId, sectionId = 1, absoluteTime = 100))
-  } else {
-    Nil
+  def buildTimeAdjustments(assessmentId: Int, application: OnlineTestApplication) = {
+    if (application.needsAdjustments || application.guaranteedInterview) {
+      List(TimeAdjustments(assessmentId, sectionId = 1, absoluteTime = calculateAbsoluteTimeWithAdjustments(application)))
```
Comment (henricook) [inline app/services/onlinetesting/Phase2TestService.scala]: What's the purpose of sectionId?

File: app/services/onlinetesting/Phase2TestService.scala [hunk 10]
```diff
@@ -205,12 +215,13 @@ trait Phase2TestService extends OnlineTestService with Phase2TestConcern with Sc
...
       resultsURL = None,
-      timeAdjustments = buildTimeAdjustments(application.needsAdjustments, schedule.assessmentId)
+      timeAdjustments = buildTimeAdjustments(schedule.assessmentId, application)
     )
   }
 
   private def insertPhase2TestGroups(o: List[Phase2TestInviteData])
-    (implicit invitationDate: DateTime, expirationDate: DateTime): Future[Unit] = Future.sequence(o.map { completedInvite =>
+                                    (implicit invitationDate: DateTime, expirationDate: DateTime): Future[Unit]
+  = Future.sequence(o.map { completedInvite =>
```
Comment (henricook) [inline app/services/onlinetesting/Phase2TestService.scala]: Same comment on the leading =

File: app/services/onlinetesting/Phase2TestService.scala [hunk 11]
```diff
@@ -356,6 +372,9 @@ trait Phase2TestService extends OnlineTestService with Phase2TestConcern with Sc
     }
   }
 
+  protected[onlinetesting] def calculateAbsoluteTimeWithAdjustments(application: OnlineTestApplication): Int = {
+    (application.eTrayAdjustments.flatMap { etrayAdjustments => etrayAdjustments.timeNeeded }.getOrElse(0) * 80 / 100) + 80
```
Comment (ollyw) [inline app/services/onlinetesting/Phase2TestService.scala]: Magic number of 80 perhaps should be a val?

File: app/repositories/onlinetesting/Phase2TestRepository.scala (hunk 2 above)
Comment (kjharkin) [inline app/repositories/onlinetesting/Phase2TestRepository.scala]: Miguel's code.

File: app/services/onlinetesting/Phase2TestService.scala (hunk 11 above)
Comment (henricook) [inline app/services/onlinetesting/Phase2TestService.scala]: Where do the 80 and 100 come from in this calculation?

File: app/services/application/ApplicationService.scala (hunk 6 above)
Comment (kjharkin) [inline app/services/application/ApplicationService.scala]: Ha ha - fair point. Will be fixed.

File: app/repositories/onlinetesting/Phase2TestRepository.scala (hunk 2 above)
Comment (ollyw) [inline app/repositories/onlinetesting/Phase2TestRepository.scala]: @mgilgar can you shed any light on the above please?

File: app/services/application/ApplicationService.scala (hunk 7 above)
Comment (kjharkin) [inline app/services/application/ApplicationService.scala]: I could argue they are but it's subjective. Removed and spacing improved.

File: app/repositories/onlinetesting/Phase2TestRepository.scala (hunk 4 above)
Comment (mgilgar) [inline app/repositories/onlinetesting/Phase2TestRepository.scala]: Done

File: app/services/onlinetesting/Phase2TestService.scala (hunk 11 above)
Comment (kjharkin) [inline app/services/onlinetesting/Phase2TestService.scala]: Miguel's code

File: app/model/Commands.scala (hunk 1 above)
Comment (kjharkin) [inline app/model/Commands.scala]: Guess not - will be changed.

File: app/repositories/onlinetesting/Phase2TestRepository.scala (hunk 5 above)
Comment (mgilgar) [inline app/repositories/onlinetesting/Phase2TestRepository.scala]: I don't think we have to verify if that works with or without.

File: app/repositories/application/GeneralApplicationRepository.scala (hunk 3 above)
Comment (kjharkin) [inline app/repositories/application/GeneralApplicationRepository.scala]: Miguel's comment

File: app/services/onlinetesting/Phase2TestService.scala (hunk 8 above)
Comment (kjharkin) [inline app/services/onlinetesting/Phase2TestService.scala]: Miguel's

File: app/services/onlinetesting/Phase2TestService.scala (hunk 10 above)
Comment (kjharkin) [inline app/services/onlinetesting/Phase2TestService.scala]: Miguel's

File: app/services/onlinetesting/Phase2TestService.scala (hunk 9 above)
Comment (kjharkin) [inline app/services/onlinetesting/Phase2TestService.scala]: Miguel's

File: app/services/onlinetesting/Phase2TestService.scala (hunk 8 above)
Comment (mgilgar) [inline app/services/onlinetesting/Phase2TestService.scala]: Yes it is weird

File: app/services/onlinetesting/Phase2TestService.scala (hunk 9 above)
Comment (mgilgar) [inline app/services/onlinetesting/Phase2TestService.scala]: It seems there are test with several sections and you can configure every section with different time.

File: app/services/onlinetesting/Phase2TestService.scala (hunk 11 above)
Comment (mgilgar) [inline app/services/onlinetesting/Phase2TestService.scala]: We pass the time extension as a percentage, so if we pass 25%, for a base duration of 80 minutes, we mean 25 * 80 / 100 = 20 extra minutes.

File: app/repositories/onlinetesting/Phase2TestRepository.scala (hunk 2 above)
Comment (ollyw) [inline app/repositories/onlinetesting/Phase2TestRepository.scala]: @mgilgar I think I have just seen the bigger picture. This actually can't be merged without https://github.com/hmrc/fset-faststream/pull/120/files. Can we discuss F2F?

Comment (mgilgar) [inline app/repositories/onlinetesting/Phase2TestRepository.scala]: @ollyw The existing code should cover that: we are not inviting those who have invigilated etray adjustment.

**Comment by ollyw** [comment]
All, can we get this merged if travis is happy. The merge window is closing, and I belief that this backend code will be dormant unless the adjustments confirmed flag is set for the scheduler, or the actual confirmation is called from the UI. I think we can manage that risk

**Comment by mgilgar** [comment]
@ollyw Then we have to make sure faststream-admin users dont apply adjustments because It seems there is a problem when we apply time extensions in cubiks.

**Pull request merged**